File: lab-solutions/Lab7/RoomPersistenceLabSkeleton/app/src/main/java/mobiledev/unb/ca/roompersistencelab/ItemRepository.java
package mobiledev.unb.ca.roompersistencelab;

import android.app.Application;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

public class ItemRepository {
    private ItemDao itemDao;

    public ItemRepository(Application application) {
        AppDatabase db = AppDatabase.getDatabase(application);
        itemDao = db.itemDao();
    }

    // TODO Add query specific methods
    //  See the example project file at
    //  https://github.com/hpowell20/cs2063-winter-2020-examples/blob/master/Lecture7/RoomPersistenceLibraryDemo/app/src/main/java/mobiledev/unb/ca/roompersistencetest/repository/ItemRepository.java
    //  to see examples of how to work with the Executor Service

    // HINTS
    //  The insert operation needs to make use of the Runnable interface
    //  The search operation needs to make use of the Callable interface

    public void insertRecord(String name, int num) {
        Item newItem = new Item();
        newItem.setName(name);
        newItem.setNum(num);

        insertRecord(newItem);
    }

    private void insertRecord(final Item item) {
        AppDatabase.databaseWriterExecutor.execute(new Runnable() {
            @Override
            public void run() {
                itemDao.insert(item);
            }
        });
    }

    /*public LiveData<List<Item>> findItemsWithName(String name) {
        return itemDao.findItemsWithName(name);
    }*/

    public List<Item> findItemsWithName(final String name) {
        Future<List<Item>> future = AppDatabase.databaseWriterExecutor.submit(new Callable<List<Item>>(){
            public List<Item> call() {
                return itemDao.findItemsWithName(name);
            }
        });

        try {
            return future.get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return new ArrayList<>();
        }
    }
}

File: Fall2022/examples/Lecture 7/RoomPersistenceLibraryDemo/app/src/main/java/mobiledev/unb/ca/roompersistencetest/repository/ItemRepository.kt
package mobiledev.unb.ca.roompersistencetest.repository

import android.app.Application
import androidx.lifecycle.LiveData
import mobiledev.unb.ca.roompersistencetest.dao.AppDatabase
import mobiledev.unb.ca.roompersistencetest.dao.AppDatabase.Companion.getDatabase
import mobiledev.unb.ca.roompersistencetest.dao.ItemDao
import mobiledev.unb.ca.roompersistencetest.entity.Item

class ItemRepository(application: Application) {
    private val itemDao: ItemDao? = getDatabase(application).itemDao()

    val allItems: LiveData<List<Item>> = itemDao!!.listAllRecords()

    fun insertRecord(name: String?, num: Int) {
        val newItem = Item()
        newItem.name = name
        newItem.num = num
        insert(newItem)
    }

    private fun insert(item: Item) {
        AppDatabase.databaseWriterExecutor.execute { itemDao!!.insert(item) }
    }

    fun deleteRecord(item: Item) {
        AppDatabase.databaseWriterExecutor.execute { itemDao!!.deleteItem(item) }
    }
}
